WIP: Tidying FCI implementation

bendudson opened the pull request:
- Fix Div_par(f, v) for FCI

bendudson committed:
difops: Replace Div_par(Field3D, Field3D)

When using FCI yup/down fields, each poloidal plane
uses a different coordinate system. Quantities like J
therefore can't be averaged between planes.

Magnetic field strength is a scalar that can be interpolated,
and is used here to calculate the divergence of a parallel flow.

bendudson committed:
Apply clang-format changes

bendudson committed:
Coordinates: Fix Div_par(f) for FCI

Use Bxy rather than J on neighboring slices: B can be compared
between slices, but J cannot (different coordinate system).

bendudson committed:
FCItransform: Only load some parallel metrics

Each poloidal slice has a different coordinate system, so metrics
can't be directly compared or averaged. Only Bxy and the parallel
connection length (dy, g_22) make sense to calculate.

bendudson committed:
DDY: Remove FCI_AUTOMAGIC calcParallelSlices

Don't calculate parallel slices in a derivative: The result is
unlikely to be correct.

bendudson committed:
Field: Extend operations to calculate in parallel slices

If the argument to `sqrt` has parallel slices, then the
slices will be operated on and the result will have parallel slices.

To avoid unnecessary work, discard slices before calling.

bendudson committed:
Field: Add withoutParallelSlices() method

Returns a shallow copy without parallel slices.
Enables user to avoid performing calculations on slices if not needed.

bendudson committed:
Apply clang-format changes

bendudson committed:
field ops: Remove FCI_AUTOMAGIC flag

Always perform calculations in the yup/down slices if present
in BOTH arguments.
i.e. automagic always on for arithmetic operators.

To avoid unnecessary work, use `withoutParallelSlices` to pass
arguments without parallel slices.

dschwoerer commented on include/bout/field.hxx:
```diff
@@ -527,22 +527,36 @@ T pow(BoutReal lhs, const T& rhs, const std::string& rgn = "RGN_ALL") {
+      /* Operate on parallel slices */                                                   \
+      result.splitParallelSlicesAndAllocate();                                           \
+      for (size_t i{0}; i != f.numberParallelSlices(); ++i) {                            \
+        BOUT_FOR(d, result.getRegion(rgn)) { result.yup(i)[d] = func(f.yup(i)[d]); }     \
```
but here we need a shifted region (at least for FCI) ...

bendudson replied:
My thinking with this was that if we are operating over a given region of a field, then we also want to operate over all the yup/down fields connected to those points. Hence the region indices should be the same for yup/down. I guess you have a different model for how this should work?

dschwoerer commented on include/bout/field3d.hxx:
```diff
@@ -237,13 +237,23 @@ public:
 
+  /// Returns a shallow copy without parallel slices
+  Field3D withoutParallelSlices() const {
+    Field3D result{getMesh(), getLocation(), getDirections(), getRegionID()};
```
Would that not be more general:
```suggestion
    auto result{emptyFrom(*this)};
```

dschwoerer commented on src/mesh/coordinates.cxx:
```diff
@@ -1600,16 +1600,14 @@ Field3D Coordinates::Div_par(const Field3D& f, CELL_LOC outloc,
-  return setName(coords->J / sqrt(coords->g_22) * Grad_par(f_B, outloc, method),
-                 "Div_par({:s})", f.name);
+  return setName(Bxy * Grad_par(f_B, outloc, method), "Div_par({:s})", f.name);
 }
```
But this about the metric coefficients, not about the strength about the metric field.
Only for Clebsch coordinates this is the same, but not for FCI.

bendudson replied:
It works because the area of a flux tube is inversely proportional to B. If we're mapping along flux tubes from one plane to the next, then the ratio of the flux tube areas is (inverse) the ratio of B on the two planes.

dschwoerer commented on src/mesh/difops.cxx:
```diff
@@ -234,57 +234,50 @@ Field3D Div_par(const Field3D& f, const std::string& method, CELL_LOC outloc) {
-      }
+  Field3D result{emptyFrom(f)};
+  BOUT_FOR(i, f.getRegion("RGN_NOBNDRY")) {
+    result[i] = B[i]
+                * ((f_up[i] * v_up[i] / B_up[i]) - (f_down[i] * v_down[i] / B_down[i]))
+                / (dy[i] * sqrt(g_22[i]));
```
This is only valid in clebsch

bendudson replied:
Ah no. This is always valid for all coordinate systems, not just Clebsch. It's just a vector identity from Div(B) = 0

Div(b v) = Div(B v / |B|) = Div(B) v/|B| + B dot Grad(v / |B|)

so the divergence of a parallel vector Div_par(v) is always B * Grad_par( v / B )

dschwoerer replied:
Because the mesh is locally field aligned - I see :+1:

dschwoerer commented on src/mesh/difops.cxx:
```diff
@@ -234,57 +234,50 @@ Field3D Div_par(const Field3D& f, const std::string& method, CELL_LOC outloc) {
+
+#if CHECK > 0
+    if (!std::isfinite(result[i])) {
+      output.write("{} {} {} {}\n", f_up[i], v_up[i], f_down[i], v_down[i]);
+      output.write("{} {} {} {} {}\n", B[i], B_up[i], B_down[i], dy[i], sqrt(g_22[i]));
```
We probably want to move that to the exception message below, and also add `i` to make it easier to figure out why that is non-finite ...

dschwoerer reviewed (commented):
I think we should talk about this, I do not think we want to remove the part about the metric coefficients?

bendudson committed:
FCI parallel: Revert deletion of coordinate slices

Added comment to explain why it's ok to use the yup/down
slices of the metric components if they are read from the grid file.

bendudson committed:
parallel_boundary_region: Fix ITER macro

Iterated for `i < localmesh->ystart - abs_offset()` so when ystart is 1
and the boundary offset is 1, no boundary is applied.

bendudson committed:
Apply clang-format changes

bendudson committed:
Fixes for FCI: Load dy, improve errors, use J in Div_par

Reverting some changes back to David's version.

bendudson committed:
Apply clang-format changes

bendudson committed:
Remove FCI_AUTOMAGIC compile option

Prefer to have consistent behavior so that code works and can be
optimized without breaking.

The field.withoutParallelSlices() method can be used to strip slices
before calling an operation, if slices are not needed.

bendudson committed:
Remove stray #endif

bendudson committed:
Field3D: Fix operator=(BoutReal)

If the field has parallel slices, assign the value also to the slices.

dschwoerer committed:
applyParallelBoundary can be a no-op for non-fci

dschwoerer committed:
Merge pull request #3202 from boutproject/fci-higher-order-tidy-noop

applyParallelBoundary can be a no-op for non-fci

dschwoerer commented:
Closing this, as the work on target branch has stopped.
